---
title: "Ancestral reconstruction"
author: "Talia_Karasov"
date: "6/17/2019"
output: html_document
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width = 12, fig.height = 8, echo = FALSE, warning = FALSE, message = FALSE, cache = TRUE)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library('dplyr')
library('treeWAS')
library(ape)
library("ggplot2")
library("phangorn")
library("phytools")
```

\
Load the phenotype data
```{r, echo = TRUE}
path = "/Users/tkarasov/work_main"
#path = "/ebio"
phen_prelim = read.csv(paste(path, "/abt6_projects8/Pseudomonas_mapping/data/mapping/emma/strain.pheno", sep=""), header = F, sep ="\t")
phen = log10(phen_prelim[,3])
names(phen) = phen_prelim[,1]
```

\
Load the tree data
```{r, echo = TRUE}
tree_file = paste(path,"/abt6_projects9/Pseudomonas_diversity/data/post_assembly_analysis/pan_genome/Ta1524/vis/strain_tree.nwk", sep="")

tree = read.tree(file = tree_file)

#good_tips = full_tree$tip.label[full_tree$tip.label %in% names(phen)]

#tree = keep.tip(full_tree, tip = good_tips)
```

\
Load the genetic data
```{r, echo = TRUE}
#This is the P/A matrix
pa_full = read.FASTA(paste(path, "/abt6_projects8/Pseudomonas_mapping/code_Pseudomonas_mapping_git/input/genePresence.fasta", sep = ""))

#the original has 72397 genes under consideration. However, subset_pa_old (only SNPs is 25860. Not clear where those 50,000 other genes are???) 

pa_convert = DNAbin2genind(pa_full)@tab

subset_pa_old = pa_convert #pa_convert[rownames(pa_convert) %in% names(phen),]

subset_pa = dplyr::select(data.frame(subset_pa_old),contains(".t"))

keep_col = subset_pa %>% 
  summarise_all(var) %>% 
  select_if(function(.) . != 0) %>% 
  names()
  
subset_pa = subset_pa[,keep_col] #12930 genes polymorphic

#out.pa <- treeWAS(snps = subset_pa, phen = phen, tree = tree, seed = 1, p.value.correct = "FDR", p.value = 0.01)

```
treeWAS does an ancestral state reconstruction. We would like to identify genes that are specific to given nodes

My node is NODE_0000046

Actually 
```{r}
#set zero length branches to 1/millionth size of tree because this function doesn't deal with zero length branches
tree$edge.length[tree$edge.length==0]<-max(nodeHeights(tree))*1e-6

keep_fit1 = list()

for(i in 1:dim(subset_pa)[2]){
  fitER<-ace(x=subset_pa[,i],tree,model="ER",type="discrete", marginal = TRUE)
  keep_fit1[i] = list(fitER)
  print(i)
}
```

Now I need to select genes that were acquired in the OTU5 node
```{r}
nodes = tree$node.label
focal_node = 110#109
ancestor_node = Siblings(tree, focal_node)
#ancestor_node = 101#getParent(tree, (focal_node)) #- Ntip(tree) #+ Ntip(tree))) 

select_my_node <- function(ace_result, focal_node = focal_node , ancestor_node = ancestor_node){
  is_there = FALSE
  if(ace_result$lik.anc[focal_node,1] >= 0.55 && ace_result$lik.anc[ancestor_node,1] <= 0.5){
    is_there = TRUE }
  else{
    is_there = FALSE }
  return(is_there[1])
}

which_genes = which(lapply(keep_fit1, select_my_node) == TRUE)
select_my_node(fitER)
```
We want to walk the tree, and for all descendant nodes, find which genes are most likely discovered in that node

```{r}
plot.phylo(tree, type = "fan", show.tip.label = FALSE, use.edge.length = TRUE)
nodelabels(pie=keep_fit1[880][[1]]$lik.anc, cex = .3, col = )

```

These results are weird because they suggest that there is no gene that was clearly gained or lost in OTU5. I think this is a probability problem with the uneven nature of the sampling


obj<-contMap(anole.tree,svl,plot=FALSE)
plot(obj,type="fan",legend=0.7*max(nodeHeights(anole.tree)),
    fsize=c(0.7,0.9))
```

FDR = 0.05
```{r, echo = TRUE}
out.pa$gene = c(1:dim(out.pa.[1]))

p1 = qplot(x = c(1:length(out.pa$subsequent$p.vals)), y = -log10(out.pa$subsequent$p.vals)) + theme_bw() + xlab("Gene Order") + ylab("-log10(p-value)")
            
p2 = qplot(x = c(1:length(out.pa$simultaneous$p.vals)), y = -log10(out.pa$simultaneous$p.vals)) + theme_bw() + xlab("Gene Order") + ylab("-log10(p-value)")

p3 = qplot(x = c(1:length(out.pa$terminal$p.vals)), y = -log10(out.pa$terminal$p.vals)) + theme_bw() + xlab("Gene Order") + ylab("-log10(p-value)")

plot_grid(p1, p2, p3, nrow = 3)
```

The second one, the simultaneous mapping, is the one that give the "significant hits"

What about if we limit the phylogenetic tree and phenotypes sampled:

```{r, echo = TRUE}
mrca = getMRCA(tree, c("p8.B3", "p24.H2"))
descendants = tree$tip.label[Descendants(tree, mrca)[[1]]]
subset_tree = keep.tip(tree, descendants)
subset_phenotype = phen[descendants]
subset_pa_new = pa_convert[rownames(pa_convert) %in% names(subset_phenotype),]
```

\ 
Run treeWAS on P/A matrix
```{r, echo = TRUE}
out.pa.subset <- treeWAS(snps = subset_pa_new, phen = subset_phenotype, tree = subset_tree, seed = 1, p.value.correct = "FDR", p.value = 0.01)
```
```
