---
title: "gene_order"
author: "Talia_Karasov"
date: "7/6/2019"
output: html_document
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width = 12, fig.height = 8, echo = FALSE, warning = FALSE, message = FALSE, cache = TRUE)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library('dplyr')
library('treeWAS')
library(ape)
library("ggplot2")
library("phangorn")
library("genbankr")
```

The goal of this script is to take the original genome and map the positions of the genes along the genome then overlay with other genomic characteristics

```{r}
suppressPackageStartupMessages(library(genbankr))

session <- ssh_connect("tkarasov@a6-karasov-lin")
#the gff reader will not read gffs with fasta sequences at the end
annotation = "~/work_main/abt6_projects8/Pseudomonas_mapping/data/mapping/references_sequences/plate25.C2.annotation_no_fasta.gff"
#smpfile = system.file(annotation, package="genbankr")
gff = read.gff(annotation)

gene_mappings = "~/work_main/abt6_projects9/Pseudomonas_diversity/data/post_assembly_analysis/pan_genome/Ta1524/sorted_genelist.txt"
gene_map = read.table(gene_mappings)

```