---
title: "TreeWAS"
author: "Talia_Karasov"
date: "6/17/2019"
output: html_document
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width = 12, fig.height = 8, echo = FALSE, warning = FALSE, message = FALSE, cache = TRUE)
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reshape2)
library('dplyr')
library('treeWAS')
library("ape")
library("ggplot2")
library("phangorn")
library("cowplot"")
```

\
Load the phenotype data
```{r, echo = TRUE}
path = "/Users/tkarasov/work_main"
phen_prelim = read.csv(paste(path, "/abt6_projects8/Pseudomonas_mapping/data/mapping/emma/strain.pheno", sep=""), header = F, sep ="\t")
phen = log10(phen_prelim[,3])
names(phen) = phen_prelim[,1]
```

\
Load the tree data
```{r, echo = TRUE}
tree_file = paste(path,"/abt6_projects9/Pseudomonas_diversity/data/post_assembly_analysis/pan_genome/Ta1524/vis/strain_tree.nwk", sep="")

full_tree = read.tree(file = tree_file)

good_tips = full_tree$tip.label[full_tree$tip.label %in% names(phen)]

tree = keep.tip(full_tree, tip = good_tips)
```

\
Load the genetic data
```{r, echo = TRUE}
#This is the P/A matrix
pa_full = read.FASTA(paste(path, "/abt6_projects8/Pseudomonas_mapping/code_Pseudomonas_mapping_git/input/genePresence.fasta", sep = ""))

pa_convert = DNAbin2genind(pa_full)@tab

subset_pa = pa_convert[rownames(pa_convert) %in% names(phen),]


```


\ 
Run treeWAS on P/A matrix
```{r, echo = TRUE}
out.pa.df <- treeWAS(snps = subset_pa, phen = phen, tree = tree, seed = 1, p.value.correct = "FDR", p.value = 0.01)
```

Now pull the "test-specified" significant associations, and find associated genes
```{r, echo = TRUE}
sig = out.pa$treeWAS.combined[1]$treeWAS.combined
sig = gsub(".t","", sig)
sig = unique(gsub(".c","", sig))

```

FDR = 0.05
```{r, echo = TRUE}
out.pa.df$gene = c(1:dim(out.pa.df[1]))

p1 = qplot(x = c(1:length(out.pa.df$subsequent$p.vals)), y = -log10(out.pa$subsequent$p.vals)) + theme_bw() + xlab("Gene Order") + ylab("-log10(p-value)")
            
p2 = qplot(x = c(1:length(out.pa$simultaneous$p.vals)), y = -log10(out.pa$simultaneous$p.vals)) + theme_bw() + xlab("Gene Order") + ylab("-log10(p-value)")

p3 = qplot(x = c(1:length(out.pa$terminal$p.vals)), y = -log10(out.pa$terminal$p.vals)) + theme_bw() + xlab("Gene Order") + ylab("-log10(p-value)")

plot_grid(p1, p2, p3, nrow = 3)
```

The second one, the simultaneous mapping, is the one that give the "significant hits"

What about if we limit the phylogenetic tree and phenotypes sampled:

```{r, echo = TRUE}
mrca = getMRCA(tree, c("p8.B3", "p24.H2"))
descendants = tree$tip.label[Descendants(tree, mrca)[[1]]]
subset_tree = keep.tip(tree, descendants)
subset_phenotype = phen[descendants]
subset_pa_new = pa_convert[rownames(pa_convert) %in% names(subset_phenotype),]
```

\ 
Run treeWAS on P/A matrix
```{r, echo = TRUE}
out.pa.subset <- treeWAS(snps = subset_pa_new, phen = subset_phenotype, tree = subset_tree, seed = 1, p.value.correct = "FDR", p.value = 0.01)
```
```
